client/authority-discovery/src/lib.rs

```diff
@@ -254,13 +254,12 @@ where
 			&self.client,
 		)?.into_iter().map(Into::into).collect::<Vec<_>>();
 
-		let signatures = key_store.read()
-			.sign_with_all(
-				key_types::AUTHORITY_DISCOVERY,
-				keys.clone(),
-				serialized_addresses.as_slice(),
-			)
-			.map_err(|_| Error::Signing)?;
+		let keystore = key_store.read();
+		let signatures = block_on(keystore.sign_with_all(
```

Comment:
I don't think blocking on async logic is an option. Instead this should properly pass a context and return `Poll::Pending` in case the future is not yet ready.

Why not use `block_on`:

Logic further below might expect to be called in a preemptive non-blocking fashion. Making that assumption in a blocking environment is problematic.

Say e.g. one executes on a single hardware thread. The logic below needs to wait for something, thus returns `Poll::Pending`. Given that one uses `block_on` here the thread is stalled and thus the dependency (the thing the logic below is waiting for) never gets any compute time. Thus one has a deadlock.

Reply:
Rather than passing a context, I think a better solution is to turn `publish_ext_addresses` into an async function and adjusting `poll` to poll it progressively.
That involves extracting `publish_ext_addresses` into a freestanding function and passing copies of `client`, `network`, etc. but I don't think that's a problem.